text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+---
+feature name: Go Language Bindings
+start date: 2020-07-17
+rfc pr: (leave this empty)
+related issue: <https://github.com/aws/aws-cdk/issues/547>
+---
+
+# Summary
+
+This RFC proposes to support Go language bindings for the CDK.
+
+# Motivation
+
+Go has seen a huge rise in popularity in recent years and is the language of choice for many infrastructure tools, particularly in the container
+world. It has ranked high in recent years on the StackOverflow Developer Survey, especially for such a young language, and with the continued
+popularity of containers and trend towards developers owning their infrastructure, there is a high demand for using Go for infrastructure-as-code
+tooling.
+
+# Design Summary
+
+In order to create the JSII Go language binding, the various types defined in the JSII spec need to be translated into equivalent types in Go. The
+three main types discussed here are Enums, Interfaces, and Classes.
+
+# Detailed Design
+
+## Enums
+
+The standard way of specifying an enum in Go is with a custom type and constant generator iota, e.g.
+
+```go
+type Month int // custom type, like type alias
+
+const (
+    Jan Month = iota + 1
+    Feb
+    Mar
+    Apr
+    May
+    ...
+)
+// Jan=1, Feb=1, Mar=3 ...
+```
+
+Since Enums in the CDK are not always iterative, using a constant generator may not be the best translation. We could instead have string values, with
+the name of the constant prefixed with the name of the Typescript enum to avoid potential namespace collisions, e.g.:
+
+```ts
+export enum LaunchType {
+  /**
+   * The service will be launched using the EC2 launch type
+   */
+  EC2 = 'EC2',
+
+  /**
+   * The service will be launched using the FARGATE launch type
+   */
+  FARGATE = 'FARGATE'
+}
+```
+
+Go translation:
+
+```go
+type LaunchType string
```

Comment:
Another addative helper utility that you might want to consider for the `enum` design for Go CDK is a method off of the LaunchType, or a function that returns the list of enumeration values. This utility would make it easier to iterate through the list of enumeration values.  (for context we recently received a feature request for this behavior in aws-sdk-go)

Reply:
Oh, interesting. Are you referring to something like this? (https://play.golang.org/p/-KyXiVS4z8s)

```go
package launchtype

type LaunchType string

const (
	LaunchTypeEc2     LaunchType = "EC2"
	LaunchTypeFargate LaunchType = "FARGATE"
)

func EC2() LaunchType {
	return LaunchTypeEc2
}

func Fargate() LaunchType {
	return LaunchTypeFargate
}

func Values() []LaunchType {
	return []LaunchType{
		LaunchTypeEc2,
		LaunchTypeFargate,
	}
}
```

Having the enum be the package name would then allow you to call it like:
```
fmt.Println(launchtype.EC2()) // => "EC2"
fmt.Println(launchtype.Fargate()) // => "FARGATE"
fmt.Println(Values()) // => [EC2 FARGATE]
```

Reply:
Similar yeah. My initial thought was a `Values` method on the `LaunchType` type. e.g.

```go
type LaunchType string

const (
     LaunchTypeEc2 LaunchType = "EC2"
     LaunchTypeFargate LaunchType = "Fargate"
)

func (LaunchType) Values() []LaunchType {
     return []LaunchType{
          LaunchTypeEc2,
          LaunchTypeFargate
     }
}
```

Wouldn't need the separate `launchtype` package for this case.

```go
fmt.Println(LaunchType("").Values())

// or
var typ  LaunchType
fmt.Println(typ.Values())
```

Alternatively could create a `LaunchTypeValues` function instead of a method, but this would have a non-zero risk of colliding with an actual enum value.